src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -310,6 +310,75 @@ class CallData {
  private:
   class QueuedPickCanceller;
 
+  class Metadata : public LoadBalancingPolicy::MetadataInterface {
+   public:
+    class Iterator : public MetadataInterface::IteratorInterface {
+     public:
+      explicit Iterator(grpc_linked_mdelem* linked_mdelem)
+          : linked_mdelem_(linked_mdelem) {}
+
+      grpc_linked_mdelem* linked_mdelem() const { return linked_mdelem_; }
+
+      virtual IteratorInterface& operator=(
+          const IteratorInterface& other) override {
+        const Iterator* it = static_cast<const Iterator*>(&other);
+        linked_mdelem_ = it->linked_mdelem_;
+        return *this;
+      }
+
+      virtual IteratorInterface& operator++() override {
+        linked_mdelem_ = linked_mdelem_->next;
+        return *this;
+      }
+
+      bool operator==(const IteratorInterface& other) const override {
+        const Iterator* it = static_cast<const Iterator*>(&other);
+        return linked_mdelem_ == it->linked_mdelem_;
+      }
+
+      virtual StringView Key() const override {
+        return StringView(GRPC_MDKEY(linked_mdelem_->md));
+      }
+
+      virtual StringView Value() const override {
+        return StringView(GRPC_MDVALUE(linked_mdelem_->md));
+      }
+
+     private:
+      grpc_linked_mdelem* linked_mdelem_;
+    };
+
+    Metadata(CallData* calld, grpc_metadata_batch* batch)
+        : calld_(calld), batch_(batch) {}
+
+    void Add(const StringView& key, const StringView& value) override {
+      grpc_linked_mdelem* linked_mdelem = static_cast<grpc_linked_mdelem*>(
+          calld_->arena_->Alloc(sizeof(grpc_linked_mdelem)));
+      bool returned_slice_is_different;
+      linked_mdelem->md = grpc_mdelem_from_slices(
+          grpc_slice_maybe_static_intern(
+              grpc_slice_from_static_buffer_internal(key.data(), key.size()),
+              &returned_slice_is_different),
+          grpc_slice_from_static_buffer_internal(value.data(), value.size()));
+      GPR_ASSERT(grpc_metadata_batch_link_tail(batch_, linked_mdelem) ==
+                 GRPC_ERROR_NONE);
+    }
+
+    IteratorInterface Begin() override { return Iterator(batch_->list.head); }
```

Comment:
I'm not sure if `Begin()` and `End()` are valid. Returning a copy of the `Iterator` as `IteratorInterface` would chop off the child fields.  I think you would need to either return a pointer or a ref.

Is the interface really needed? I personally find virtual `Being()` and `End()` methods a little bit strange.

Reply:
I see. Another option to avoid a dynamic allocation is have the `Iterator` as a final class in `MetdataInterface`, and then having the `Iterator` call a private abstract method from  `MetdataInterface` which returns `next` based on the current `iter`. For instance, the interface can provide `Size()` and `At(size index, string_view& key, string_view& val)`.

Reply:
An interface using `Size()` and `At()` assumes that the implementation is keeping the values in an array.  If possible, I'd rather have a more abstract interface that allows the implementation to choose its underlying data structure -- e.g., it should also work if the metadata is stored in a map.

That having been said, I don't yet have a better idea, so I might wind up doing this anyway.  But let me see if I can think of a better alternative first.

Reply:
Another random idea to be able to support different containers:

Most iterators can fit in a pointer, so if we add a `void*` scratch space to the `Iterator` final class, the implementations of `MetadataInterface` should be able to handle the details on their own.  If some implementation needs more space, they would have to allocate their internal iterator data on the heap, but for `Metadata` there would be no allocation. 

```
class MDInterface {
 public:
  class Iterator {
    public:
      StringView key() const { return md_->GetKey(scratch_); }
      StringView val() const { return md_->GetValue(scratch_); }

     ... operator++() {
       scratch_ = md_->Next(scratch_);
    }

    private:
      MetadataInterface* md_;
      void* scratch_;
      friend class MDInterface;
  }

 protected:
   Iterator CreateIter(void* scratch) { 
       return {this, scratch};
   }
};
```

Next(const Iterator& iter)

Reply:
It's a great idea. Given the constraints here, I think it's reall neat. :-)